File: packages/flutter-im-ui/lib/src/models/form_behavior.dart
enum FlareFormControlState {
  idle,
  hover,
  focus,
  pressed,
  filled,
  invalid,
  disabled,
  readOnly,
  loading,
}

enum FlareFormKeyboardIntent {
  none,
  tabForward,
  tabBackward,
  submit,
  closePopup,
  nextOption,
  previousOption,
  firstOption,
  lastOption,
  pageForward,
  pageBackward,
}

FlareFormKeyboardIntent resolveFormKeyboardIntent(
  String key, {
  bool shift = false,
  bool composing = false,
  bool popupOpen = false,
  bool submitOnEnter = false,
}) {
  // While an IME composition is open the IME owns the keyboard — Enter commits a
  // candidate, Escape cancels the composition, Tab and the arrows walk the
  // candidate window. Composing short-circuits everything.
  if (composing) return FlareFormKeyboardIntent.none;
  final normalized = key.toLowerCase();
  if (normalized == 'tab')
    return shift
        ? FlareFormKeyboardIntent.tabBackward
        : FlareFormKeyboardIntent.tabForward;
  if (normalized == 'escape' && popupOpen)
    return FlareFormKeyboardIntent.closePopup;
  if (normalized == 'enter' && submitOnEnter)
    return FlareFormKeyboardIntent.submit;
  if (!popupOpen) return FlareFormKeyboardIntent.none;
  return switch (normalized) {
    'arrowdown' => FlareFormKeyboardIntent.nextOption,
    'arrowup' => FlareFormKeyboardIntent.previousOption,
    'home' => FlareFormKeyboardIntent.firstOption,
    'end' => FlareFormKeyboardIntent.lastOption,
    'pagedown' => FlareFormKeyboardIntent.pageForward,
    'pageup' => FlareFormKeyboardIntent.pageBackward,
    _ => FlareFormKeyboardIntent.none,
  };
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/FormBehavior.kt
package com.flare.im.ui

enum class FlareFormControlState { Idle, Hover, Focus, Pressed, Filled, Invalid, Disabled, ReadOnly, Loading }
enum class FlareFormKeyboardIntent { None, TabForward, TabBackward, Submit, ClosePopup, NextOption, PreviousOption, FirstOption, LastOption, PageForward, PageBackward }

fun resolveFormKeyboardIntent(
    key: String,
    shift: Boolean = false,
    composing: Boolean = false,
    popupOpen: Boolean = false,
    submitOnEnter: Boolean = false,
): FlareFormKeyboardIntent {
    // While an IME composition is open the IME owns the keyboard — Enter commits a
    // candidate, Escape cancels the composition, Tab and the arrows walk the
    // candidate window. Composing short-circuits everything.
    if (composing) return FlareFormKeyboardIntent.None
    val normalized = key.lowercase()
    if (normalized == "tab") return if (shift) FlareFormKeyboardIntent.TabBackward else FlareFormKeyboardIntent.TabForward
    if (normalized == "escape" && popupOpen) return FlareFormKeyboardIntent.ClosePopup
    if (normalized == "enter" && submitOnEnter) return FlareFormKeyboardIntent.Submit
    if (!popupOpen) return FlareFormKeyboardIntent.None
    return when (normalized) {
        "arrowdown" -> FlareFormKeyboardIntent.NextOption
        "arrowup" -> FlareFormKeyboardIntent.PreviousOption
        "home" -> FlareFormKeyboardIntent.FirstOption
        "end" -> FlareFormKeyboardIntent.LastOption
        "pagedown" -> FlareFormKeyboardIntent.PageForward
        "pageup" -> FlareFormKeyboardIntent.PageBackward
        else -> FlareFormKeyboardIntent.None
    }
}

/**
 * The text a capped field keeps when the user types past [maxLength].
 *
 * Clamping keeps the field and the state in sync. Dropping the update — what the
 * composer and the input used to do — leaves the IME composing against a value
 * the state no longer holds, and CJK input silently loses characters at the
 * boundary. `null` means no cap.
 */
fun composerTextWithinLimit(next: String, maxLength: Int?): String =
    if (maxLength != null && next.length > maxLength) next.take(maxLength) else next
